### PR TITLE
ipc: component free fix

### LINKED ISSUE #3012: [BUG] SOF_IPC_PANIC_EXCEPTION when CONFIG_DEBUG_BLOCK_FREE is enabled
**Describe the bug**
SOF_IPC_PANIC_EXCEPTION occurs when compiling FW with CONFIG_DEBUG_BLOCK_FREE = y.
Panic happens after sending TplgCompFree ipc. 

**To Reproduce**
Run slim driver python tests:
eg. TestPtpSimpleLoopback48000Hz32b32b2ch from fw_02_pass_through package.

**Reproduction Rate**
100%

**Expected behavior**
Panic shouldn't occurred.

**Environment**
I checked with WHL RVP.
Commit: 5d190901e045041d0a735ac2a7028cf09419a813 SOF master branch

**Logs**
[dma_trace_1.log](https://github.com/thesofproject/sof/files/4717054/dma_trace_1.log)
[dma_trace_1_panic_full.log](https://github.com/thesofproject/sof/files/4717055/dma_trace_1_panic_full.log)
[DECODED_COREDUMP.txt](https://github.com/thesofproject/sof/files/4717056/DECODED_COREDUMP.txt)

[02_00_TestPtpSimpleLoopback48000Hz32b32b2ch.zip](https://github.com/thesofproject/sof/files/4717058/02_00_TestPtpSimpleLoopback48000Hz32b32b2ch.zip)

### PR DESCRIPTION
**Pull request by abonislawski**
This will allow to free component properly and stop using
already freed memory which can result in dsp panic if configured
with CONFIG_DEBUG_BLOCK_FREE

Fix in comp_free() is obvious but second one in ipc_buffer_free() is more tricky.
Buffer component hold pointers to sink and source component which may be already freed.
In this patch ipc_buffer_free() will check if sink and source pointers are still available in component list:
if yes - they are ok and can be accessed to check state,
if not - they are already freed and we can free buffer

Please comment if there is a better way to check this

Fixes #3012

### CODE REVIEW

File: src/ipc/ipc.c [hunk 1]
```diff
@@ -309,10 +311,20 @@ int ipc_buffer_free(struct ipc *ipc, uint32_t buffer_id)
 	if (!cpu_is_me(ibd->core))
 		return ipc_process_on_core(ibd->core);
 
-	/* check state */
-	if (ibd->cb->sink->state != COMP_STATE_READY ||
-	    ibd->cb->source->state != COMP_STATE_READY)
-		return -EINVAL;
+	/* for each component */
```
Comment (lgirdwood) [inline src/ipc/ipc.c:314]: what are we doing here for each componet ? can you say in the comment.

Comment (abonislawski) [reply src/ipc/ipc.c]: comment updated

File: src/ipc/ipc.c [hunk 2]
```diff
@@ -309,10 +311,20 @@ int ipc_buffer_free(struct ipc *ipc, uint32_t buffer_id)
 	if (!cpu_is_me(ibd->core))
 		return ipc_process_on_core(ibd->core);
 
-	/* check state */
-	if (ibd->cb->sink->state != COMP_STATE_READY ||
-	    ibd->cb->source->state != COMP_STATE_READY)
-		return -EINVAL;
+	/* try to find sink/source components */
```
Comment (lgirdwood) [inline src/ipc/ipc.c:314]: this is better, but I want to know why we are doing this check. What condition are we checking is true before we free ?

Comment (abonislawski) [reply src/ipc/ipc.c]: Updated, note that it is divided into 2 comments (another one inside this loop)

**Comment by lgirdwood** [comment]
Jenkins seems to have stalled - I can ignore the build problem on CHT atm as this can be easily fixed in parallel.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by marc-hb** [comment]
Yet another green failure, cht + xcc in https://sof-ci.01.org/sofpr/PR3060/build6354/build/  claims SUCCESS yet in https://sof-ci.01.org/sofpr/PR3060/build6354/build/cht_xcc.txt
```
[100%] Linking C executable sof
/home/test/xcc/install/tools/RD-2012.5-linux/XtensaTools/bin/xt-ld:
    sof section `.text' will not fit in region `sof_text_start'
CMakeFiles/sof.dir/build.make:1820: recipe for target 'sof' failed
make[3]: *** [sof] Error 2
```

Thanks @fredoh9 for spotting this.

**Comment by lgirdwood** [comment]
Jenkins has been reset rerun tests.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Pull request merged**